api/mfx_dispatch/windows/include/mfx_dispatcher_log.h

```diff
@@ -0,0 +1,282 @@
+#include <string>
+
+template <class T> 
+class DSSingleTone
+{
+public:
+    template <class TParam1>
+    inline static T & get(TParam1 par1)
+    {
+        T * pstored;
+        if (NULL == (pstored = store_or_load()))
+        {
+            return *store_or_load(new T(par1));
+        }
+        return *pstored;    
+    }
+
+    inline static T & get()
+    {
+        T * pstored;
+        if (NULL == (pstored = store_or_load()))
+        {
+            return *store_or_load(new T());
+        }
+        return *pstored;
+    }
+private:
+    //if obj == NULL, then it load 
+    //if obj != NULL then it store obj
+    inline static T * store_or_load(T * obj = NULL)
+    {
+        static std::unique_ptr<T> instance;
+        if (NULL != obj)
+        {
+            instance.reset(obj);
+        }
+        return instance.get();
+    }
+};
+
+class DispatchLog 
+    : public DSSingleTone<DispatchLog>
+{
+    friend class DSSingleTone<DispatchLog>;
+    std::list<IMsgHandler*>m_Recepients;
+    int m_DispatcherLogSink;
+
+public:
+    //sets current sink
+    void   SetSink(int nsink, IMsgHandler *pHandler);
+    void   AttachSink(int nsink, IMsgHandler *pHandler);
+    void   DetachSink(int nsink, IMsgHandler *pHandler);
+    void   ExchangeSink(int nsink, IMsgHandler *pOld, IMsgHandler *pNew);
+    void   DetachAllSinks();
+    void   Write(int level, int opcode, const char * msg, va_list argptr);
+    
+protected:
+    DispatchLog();
+};
+
+//allows to push arguments on the stack without declaring them as function parameters
+struct  DispatcherLogBracketsHelper
+{
+    int  m_level;
+    int  m_opcode;
+    DispatcherLogBracketsHelper(int level, int opcode)
+        :m_level(level)
+        ,m_opcode(opcode)
+    {
+    }
+    void Write(const char * str, ...);
+} ;
+
+//auto log on ctor dtor
+struct DispatchLogBlockHelper
+{
+    int  m_level;
+    void Write(const char * str, ...);
+    DispatchLogBlockHelper (int level)
+        : m_level(level)
+    {
+    }
+    ~DispatchLogBlockHelper();
+};
+
+//----utility sinks-----
+#if defined(DISPATCHER_LOG_REGISTER_EVENT_PROVIDER)
+class ETWHandlerFactory
+    : public DSSingleTone<ETWHandlerFactory>
+{
+    friend class DSSingleTone<ETWHandlerFactory>;
+    typedef std::map<std::wstring, IMsgHandler*> _storage_type;
+    _storage_type m_storage;
+
+public:
+    ~ETWHandlerFactory();
+    IMsgHandler *GetSink(const wchar_t* sguid = DISPATCHER_LOG_EVENT_GUID);
+
+protected:
+    ETWHandlerFactory(){}
+};
+#endif
+
+#if defined(DISPATCHER_LOG_REGISTER_FILE_WRITER)
+class FileSink 
+    : public DSSingleTone<FileSink>
+    , public IMsgHandler
+{
+    friend class DSSingleTone<FileSink>;
+public:
+    virtual void Write(int level, int opcode, const char * msg, va_list argptr);
+    ~FileSink()
+    {
+        if (NULL != m_hdl)
+            fclose(m_hdl);
+    }
+private:
+    FILE * m_hdl;
+    FileSink(const std::string & log_file)
+    {
+        fopen_s(&m_hdl, log_file.c_str(), "a");
```

Comment:
on linux we don't use '_s' functions. I know that it is windows code, but maybe better to align this. @dvrogozh what do you think?

Reply:
Should be OK for Windows

Reply:
On Linux there is no fopen_s, but this is Windows-only code. So, you can use fopen_s in this code. However, if you are doing this, do that correctly. According to https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/fopen-s-wfopen-s?view=vs-2017, function prototype is:
```
errno_t fopen_s(
   FILE** pFile,
   const char *filename,
   const char *mode
);
```
You miss error handling. Probably you need to check it and throw exception.

Reply:
As I understand this logic was introduced by some reason: it will continue execution, even call of fopen_s was failed (FileSink::Write handles case when m_hdl==NULL). Documentation says content of pFile will be unchanged (==NULL) in case of error.
I can only suggest explicitly set m_hdl=NULL before call of fopen_s.

Reply:
'file' is private field, left not initialized by default and we discuss constructor. Looks like if current execution will continue it will be undefined behavior.